// next-util/index.js
'use strict';

const yaml = require('js-yaml');
const fs = require('fs');
const path = require('path');
const { merge } = require('lodash');

module.exports = function(hexo, pluginDir) {
  hexo.extend.filter.register('before_generate', function err() {
    hexo.log.warn(`The upstream repository of theme NexT has been changed.`);
    hexo.log.warn(`For more information: https://github.com/next-theme/hexo-theme-next`);
    hexo.log.warn(`Documentation: https://theme-next.js.org`);
  });
  this.hexo = hexo;
  this.pluginDir = pluginDir;
  this.getFilePath = function(file) {
    return this.pluginDir ? path.resolve(this.pluginDir, file) : file;
  };
  this.getFileContent = function(file) {
    return fs.readFileSync(this.getFilePath(file), 'utf8');
  };
  this.defaultConfigFile = function(key, file) {
    const defaultConfig = file ? yaml.safeLoad(this.getFileContent(file)) : {};
    this.hexo.config[key] = merge(defaultConfig[key], this.hexo.theme.config[key], this.hexo.config[key]);
    return this.hexo.config[key];
  };
};
